enable max-complexity gate

We currently have flake8 `max-complexity` disabled because it would break the flake8 gate. This is because the `main` function in `hubble.shell` is absolutely absurdly complicated:

```
./hubble/shell.py:240:1: C901 'main' is too complex (24)
```

We should refactor that function in to something less stupid, and enable the `max-complexity` gate.

--- a/hubble/config.py
+++ b/hubble/config.py
@@ -14,7 +14,6 @@
 
 from configparser import _UNSET, \
     NoOptionError, NoSectionError, RawConfigParser
-
 from io import StringIO
 from itertools import chain
 import os

--- a/hubble/config.py
+++ b/hubble/config.py
@@ -12,11 +12,13 @@
 #   See the License for the specific language governing permissions and
 #   limitations under the License.
 
+from configparser import _UNSET, \
+    NoOptionError, NoSectionError, RawConfigParser
+
+from io import StringIO
 from itertools import chain
 import os
 
-from backports.configparser import _UNSET, \
-    NoOptionError, NoSectionError, RawConfigParser
 from six import string_types
 
 
@@ -102,9 +104,18 @@ class ErrorConfigParser(SafeConfigParser):
 def open_fd(file):
     """ Open the file if possible, else return None """
     try:
+        if isinstance(file, StringIO):
+            return file
         return open(file)
     except IOError:
         return None
+
+
+def exists(obj):
+    """Returns true if obj is a StringIO or if the path exists """
+    if isinstance(obj, StringIO):
+        return True
+    return os.pathy.exists(obj)
 
 
 def read_configs(files=None, default_section=None):
@@ -114,7 +125,7 @@ def read_configs(files=None, default_section=None):
     """
     files = files or [os.path.expanduser('~/.hubblerc'), '.hubblerc']
     # If non of these files exist, raise an error
-    if not any([os.path.exists(rc) for rc in files]):
+    if not any([exists(rc) for rc in files]):
         return ErrorConfigParser("Unable to find config files in these"
                                  " locations [%s]" % ", ".join(files))
     return parse_configs([open_fd(file) for file in files], default_section)

--- a/hubble/config.py
+++ b/hubble/config.py
@@ -103,9 +103,9 @@ class ErrorConfigParser(SafeConfigParser):
 
 def open_fd(file):
     """ Open the file if possible, else return None """
+    if not isinstance(file, string_types):
+        return file
     try:
-        if isinstance(file, StringIO):
-            return file
         return open(file)
     except IOError:
         return None
@@ -115,7 +115,7 @@ def exists(obj):
     """Returns true if obj is a StringIO or if the path exists """
     if isinstance(obj, StringIO):
         return True
-    return os.pathy.exists(obj)
+    return os.path.exists(obj)
 
 
 def read_configs(files=None, default_section=None):

--- a/hubble/config.py
+++ b/hubble/config.py
@@ -12,12 +12,12 @@
 #   See the License for the specific language governing permissions and
 #   limitations under the License.
 
-from configparser import _UNSET, \
-    NoOptionError, NoSectionError, RawConfigParser
 from io import StringIO
 from itertools import chain
 import os
 
+from backports.configparser import _UNSET, \
+    NoOptionError, NoSectionError, RawConfigParser
 from six import string_types
 
 

--- a/hubble/shell.py
+++ b/hubble/shell.py
@@ -18,7 +18,7 @@ import argparse
 import logging
 import os
 import re
-from subprocess import CalledProcessError, check_output, PIPE, Popen
+from subprocess import check_output, PIPE, Popen
 import sys
 import textwrap
 
@@ -193,7 +193,7 @@ def get_environments(args, choice, config):
         # if -o was passed on the commandline
         if args.option:
             if 'opt-cmd' not in env:
-                log.warning("provided -o|--option, but 'opt-cmd' is not " \
+                log.warning("provided -o|--option, but 'opt-cmd' is not "
                             "defined in '%s' section" % env['section'].value)
             else:
                 env.add(run(env['opt-cmd'].value, env))

--- a/hubble/shell.py
+++ b/hubble/shell.py
@@ -270,6 +270,7 @@ def eval_args(conf, parser):
 
 def execute_environment(conf, env, hubble_args, other_args):
     cmd = get_cmd(conf, env, hubble_args)
+    cmd = cmd + other_args
 
     # If --debug; print out our env config and pass along the
     # --debug arg
@@ -287,7 +288,7 @@ def execute_environment(conf, env, hubble_args, other_args):
 
     try:
         # Run the requested command
-        p = Popen([env['cmd'].value] + other_args,
+        p = Popen(cmd,
                   stdout=PIPE,
                   stderr=PIPE,
                   env=environ)

--- a/hubble/shell.py
+++ b/hubble/shell.py
@@ -188,6 +188,8 @@ def get_environments(args, choice, config):
         # Add the args to the environment as opt.'<arg_name>'
         env.add(dict(map(f, vars(args).items())), section)
 
+        env.eval()
+
         # Populate environment vars by running opt-cmd
         # if -o was passed on the commandline
         if 'opt-cmd' in env:
@@ -198,7 +200,7 @@ def get_environments(args, choice, config):
             env.add(run(env['env-cmd'].value, env))
 
         # Apply var expansion
-        results.append(env.eval())
+        results.append(env)
     return results
 
 

--- a/hubble/shell.py
+++ b/hubble/shell.py
@@ -14,17 +14,17 @@
 
 from __future__ import print_function
 
+from hubble.config import read_configs
+
+from configparser import NoOptionError, NoSectionError
+from subprocess import check_output, PIPE, Popen
 import argparse
+import textwrap
 import logging
+import sys
 import os
 import re
-from subprocess import check_output, PIPE, Popen
-import sys
-import textwrap
 
-from backports.configparser import NoOptionError, NoSectionError
-
-from hubble.config import read_configs
 
 try:
     # Not everyone needs keyring
@@ -78,7 +78,7 @@ class Env(dict):
                 # Delete the key from the env
                 self.delete(key)
             # else add it
-            self.set(key, value, section, key.isupper())
+            self.set(key, value, section)
 
     def eval(self):
         """ Exapand all the ${variable} directives in the collection """
@@ -146,10 +146,10 @@ def green(msg):
     return "\033[92m%s\033[0m" % msg
 
 
-def get_cmd(conf, env, hubble_args):
+def get_cmd(argv, conf, env, hubble_args):
     # If our invocation name is not 'hubble'
-    if not sys.argv[0].endswith('hubble'):
-        return cmd_path(sys.argv[0], conf)
+    if not argv[0].endswith('hubble'):
+        return cmd_path(argv[0], conf)
 
     if hubble_args.execute:
         # Use the command provided
@@ -188,15 +188,6 @@ def get_environments(args, choice, config):
             return "opt.%s" % i[0], str(i[1])
         # Add the args to the environment as opt.'<arg_name>'
         env.add(dict(map(f, vars(args).items())), section)
-
-        # Populate environment vars by running opt-cmd
-        # if -o was passed on the commandline
-        if 'opt-cmd' in env:
-            env.add(run(env['opt-cmd'].value, env))
-
-        # Populate environment vars by running the env-cmd if it exists
-        if 'env-cmd' in env:
-            env.add(run(env['env-cmd'].value, env))
 
         # Apply var expansion
         results.append(env.eval())
@@ -249,7 +240,7 @@ def cmd_path(cmd, conf):
     return "/usr/bin/%s" % basename
 
 
-def eval_args(conf, parser):
+def eval_args(argv, conf, parser):
     env = conf.safe_get(conf.default_section, 'default-env')
     # If no default environment set, look for an
     # environment choice on the command line
@@ -257,23 +248,34 @@ def eval_args(conf, parser):
         help = "The environment defined in ~/.hubblerc to use"
         parser.add_argument('env', nargs='?', metavar='<ENV>',
                             help=help)
-        (arg1, arg2) = parser.parse_known_args()
-        return (arg1, arg2, arg1.env)
+        (arg1, arg2) = parser.parse_known_args(args=argv[1:])
+        return arg1, arg2, arg1.env
     # Return the args with the default environment choice
-    (arg1, arg2) = parser.parse_known_args()
-    return (arg1, arg2, env)
+    (arg1, arg2) = parser.parse_known_args(args=argv[1:])
+    return arg1, arg2, env
 
 
-def execute_environment(conf, env, hubble_args, other_args):
-    cmd = get_cmd(conf, env, hubble_args)
-    cmd = cmd + other_args
+def execute_environment(cmd, env, hubble_args, other_args):
+    # Populate environment vars by running opt-cmd
+    # if -o was passed on the commandline
+    if hubble_args.option:
+        if 'opt-cmd' not in env:
+            log.warning("provided -o|--option, but 'opt-cmd' is not "
+                        "defined in '%s' section" % env['section'].value)
+        else:
+            env.add(run(env['opt-cmd'].value, env))
+
+    # Populate environment vars by running the env-cmd if it exists
+    if 'env-cmd' in env:
+        env.add(run(env['env-cmd'].value, env))
 
     # If --debug; print out our env config and pass along the
     # --debug arg
     if hubble_args.debug:
         # For cinder client debug
-        if cmd.endswith('cinder'):
+        if cmd.value.endswith('cinder'):
             env.add({'CINDERCLIENT_DEBUG': '1'})
+        env.add({'cmd': cmd.value})
         print("%r\n" % env)
         other_args.insert(0, '--debug')
 
@@ -282,9 +284,10 @@ def execute_environment(conf, env, hubble_args, other_args):
     environ = os.environ.copy()
     environ.update(env.to_dict())
 
+    print(environ)
     try:
         # Run the requested command
-        p = Popen(cmd,
+        p = Popen([cmd.value] + other_args,
                   stdout=PIPE,
                   stderr=PIPE,
                   env=environ)
@@ -292,14 +295,14 @@ def execute_environment(conf, env, hubble_args, other_args):
         if e.errno == 2:
             print("-- No such executable '%s', you must specify the "
                   "executable in the [hubble-commands] section of the "
-                  "config (See README)" % env['cmd'].value)
-        message = "exec failed '%s' - %s" % (env['cmd'].value, e)
+                  "config (See README)" % cmd.value)
+        message = "exec failed '%s' - %s" % (cmd.value, e)
         raise RuntimeError(message)
 
     return p
 
 
-def main():
+def main(argv=sys.argv, stdout=sys.stdout, stderr=sys.stderr, files=None):
     logging.basicConfig(format='-- %(message)s')
     log.setLevel(logging.CRITICAL)
     formatter_class = argparse.RawDescriptionHelpFormatter
@@ -311,7 +314,7 @@ def main():
             on environment variables for configuration.
 
             Use ~/.hubblerc for user wide environments then place a
-            .hubblerc in a local directory to overide ~/.hubblerc
+            .hubblerc in a local directory to override ~/.hubblerc
             """))
     parser.add_argument('-o', '--option',
                         help="an argument to pass to the opt-cmd")
@@ -324,10 +327,10 @@ def main():
                         "and passes --debug to selected command")
 
     # Read the configs
-    conf = read_configs(default_section='hubble')
+    conf = read_configs(files=files, default_section='hubble')
     # Evaluate the command line arguments and return our args
     # the commands args and the environment choice the user made
-    hubble_args, other_args, choice = eval_args(conf, parser)
+    hubble_args, other_args, choice = eval_args(argv, conf, parser)
     # Do this so we pass along the -h to the command
     # if we are using invocation discovery
     if hubble_args.help and (choice is None):
@@ -349,17 +352,21 @@ def main():
     if hubble_args.debug:
         log.setLevel(logging.DEBUG)
 
-    # Read environment values from config files
+    # Collect all environments from our config file
     environments = get_environments(hubble_args, choice, conf)
     processes = []
     for env in environments:
-        p = execute_environment(conf, env, hubble_args, other_args)
+        # Get the command to execute
+        cmd = get_cmd(argv, conf, env, hubble_args)
+        # Create the selected environment to execute our command in
+        p = execute_environment(cmd, env, hubble_args, other_args)
         processes.append((p, env['section'].value))
 
     for p, env in processes:
         if len(environments) != 1:
             print("-- [%s] --" % green(env))
         # Wait for the command to complete
-        stdout, stderr = p.communicate()
-        sys.stdout.write(stdout.decode('utf-8'))
-        sys.stderr.write(stderr.decode('utf-8'))
+        out, err = p.communicate()
+        stdout.write(out.decode('utf-8'))
+        stderr.write(err.decode('utf-8'))
+    return 0

--- a/hubble/shell.py
+++ b/hubble/shell.py
@@ -15,13 +15,14 @@
 from __future__ import print_function
 
 import argparse
-from configparser import NoOptionError, NoSectionError
 import logging
 import os
 import re
 from subprocess import check_output, PIPE, Popen
 import sys
 import textwrap
+
+from backports.configparser import NoOptionError, NoSectionError
 
 from hubble.config import read_configs
 

--- a/hubble/shell.py
+++ b/hubble/shell.py
@@ -14,17 +14,16 @@
 
 from __future__ import print_function
 
-from hubble.config import read_configs
-
-from configparser import NoOptionError, NoSectionError
-from subprocess import check_output, PIPE, Popen
 import argparse
-import textwrap
+from configparser import NoOptionError, NoSectionError
 import logging
-import sys
 import os
 import re
+from subprocess import check_output, PIPE, Popen
+import sys
+import textwrap
 
+from hubble.config import read_configs
 
 try:
     # Not everyone needs keyring

--- a/hubble/shell.py
+++ b/hubble/shell.py
@@ -281,7 +281,6 @@ def execute_environment(cmd, env, hubble_args, other_args):
     environ = os.environ.copy()
     environ.update(env.to_dict())
 
-    print(environ)
     try:
         # Run the requested command
         p = Popen([cmd.value] + other_args,

--- a/hubble/shell.py
+++ b/hubble/shell.py
@@ -188,6 +188,15 @@ def get_environments(args, choice, config):
         # Add the args to the environment as opt.'<arg_name>'
         env.add(dict(map(f, vars(args).items())), section)
 
+        # Populate environment vars by running opt-cmd
+        # if -o was passed on the commandline
+        if 'opt-cmd' in env:
+            env.add(run(env['opt-cmd'].value, env))
+
+        # Populate environment vars by running the env-cmd if it exists
+        if 'env-cmd' in env:
+            env.add(run(env['env-cmd'].value, env))
+
         # Apply var expansion
         results.append(env.eval())
     return results
@@ -255,19 +264,6 @@ def eval_args(argv, conf, parser):
 
 
 def execute_environment(cmd, env, hubble_args, other_args):
-    # Populate environment vars by running opt-cmd
-    # if -o was passed on the commandline
-    if hubble_args.option:
-        if 'opt-cmd' not in env:
-            log.warning("provided -o|--option, but 'opt-cmd' is not "
-                        "defined in '%s' section" % env['section'].value)
-        else:
-            env.add(run(env['opt-cmd'].value, env))
-
-    # Populate environment vars by running the env-cmd if it exists
-    if 'env-cmd' in env:
-        env.add(run(env['env-cmd'].value, env))
-
     # If --debug; print out our env config and pass along the
     # --debug arg
     if hubble_args.debug:

--- a/hubble/shell.py
+++ b/hubble/shell.py
@@ -262,109 +262,100 @@ def main():
                         help="Adds CINDERCLIENT_DEBUG=1 to the environment "
                         "and passes --debug to selected command")
 
-    try:
-        # Read the configs
-        conf = read_configs(default_section='hubble')
-        # Evaluate the command line arguments and return our args
-        # the commands args and the environment choice the user made
-        hubble_args, other_args, choice = eval_args(conf, parser)
-        # Do this so we pass along the -h to the command
-        # if we are using invocation discovery
-        if hubble_args.help and (choice is None):
-            return parser.print_help()
+    # Read the configs
+    conf = read_configs(default_section='hubble')
+    # Evaluate the command line arguments and return our args
+    # the commands args and the environment choice the user made
+    hubble_args, other_args, choice = eval_args(conf, parser)
+    # Do this so we pass along the -h to the command
+    # if we are using invocation discovery
+    if hubble_args.help and (choice is None):
+        return parser.print_help()
 
-        # If there was an error
-        if conf.get_error():
-            print(conf.get_error())
-            return 1
+    # If there was an error
+    if conf.get_error():
+        print(conf.get_error())
+        return 1
 
-        if choice is None:
-            print("Environments Configured: %s" % ",".join(conf.sections()))
-            print("See --help for usage")
-            return 1
-        if hubble_args.help:
-            other_args.append('--help')
+    if choice is None:
+        print("Environments Configured: %s" % ",".join(conf.sections()))
+        print("See --help for usage")
+        return 1
+    if hubble_args.help:
+        other_args.append('--help')
 
-        # Set our log level
-        if hubble_args.debug:
-            log.setLevel(logging.DEBUG)
+    # Set our log level
+    if hubble_args.debug:
+        log.setLevel(logging.DEBUG)
 
-        # Read environment values from config files
-        environments = get_environments(hubble_args, choice, conf)
-        processes = []
-        for env in environments:
-            # Populate environment vars by running opt-cmd
-            # if -o was passed on the commandline
-            if hubble_args.option:
-                if 'opt-cmd' not in env:
-                    message = "provided -o|--option, but 'opt-cmd' is not " \
-                              "defined in '%s' section" % env['section'].value
-                    raise RuntimeError(message)
-                env.add(run(env['opt-cmd'].value, env))
-
-            # Populate environment vars by running the env-cmd if it exists
-            if 'env-cmd' in env:
-                env.add(run(env['env-cmd'].value, env))
-
-            # If querying multiple environments, display the env name
-            if hubble_args.debug:
-                print("-- [%s] --" % green(env['section'].value))
-
-            # If our invocation name is not 'hubble'
-            if not sys.argv[0].endswith('hubble'):
-                # Use the invocation name as our 'cmd'
-                env.add({'cmd': cmd_path(sys.argv[0], conf)})
-
-            if hubble_args.execute:
-                # Use the command provided
-                env.add({'cmd': hubble_args.execute})
-
-            # At this point we should know our 'cmd' to run
-            if 'cmd' not in env:
-                raise RuntimeError("Please specify a 'cmd' somewhere in "
-                                   "your config")
-
-            # If --debug; print out our env config and pass along the
-            # --debug arg
-            if hubble_args.debug:
-                # For cinder client debug
-                if env['cmd'].endswith('cinder'):
-                    env.add({'CINDERCLIENT_DEBUG': '1'})
-                print("%r\n" % env)
-                other_args.insert(0, '--debug')
-
-            # Grab a copy of the local environment and inject it into
-            # our environment
-            environ = os.environ.copy()
-            environ.update(env.to_dict())
-
-            try:
-                # Run the requested command
-                p = Popen([env['cmd'].value] + other_args,
-                          stdout=PIPE,
-                          stderr=PIPE,
-                          env=environ)
-                processes.append((p, env['section'].value))
-            except OSError as e:
-                if e.errno == 2:
-                    print("-- No such executable '%s', you must specify the "
-                          "executable in the [hubble-commands] section of the "
-                          "config (See README)" % env['cmd'].value)
-                message = "exec failed '%s' - %s" % (env['cmd'].value, e)
+    # Read environment values from config files
+    environments = get_environments(hubble_args, choice, conf)
+    processes = []
+    for env in environments:
+        # Populate environment vars by running opt-cmd
+        # if -o was passed on the commandline
+        if hubble_args.option:
+            if 'opt-cmd' not in env:
+                message = "provided -o|--option, but 'opt-cmd' is not " \
+                          "defined in '%s' section" % env['section'].value
                 raise RuntimeError(message)
+            env.add(run(env['opt-cmd'].value, env))
 
-        for p, env in processes:
-            if len(environments) != 1:
-                print("-- [%s] --" % green(env))
-            # Wait for the command to complete
-            stdout, stderr = p.communicate()
-            sys.stdout.write(stdout.decode('utf-8'))
-            sys.stderr.write(stderr.decode('utf-8'))
+        # Populate environment vars by running the env-cmd if it exists
+        if 'env-cmd' in env:
+            env.add(run(env['env-cmd'].value, env))
 
-    except (RuntimeError, NoSectionError) as e:
-        log.critical(e)
-        return 1
-    except CalledProcessError as e:
-        log.critical(e.output)
-        log.critical(e)
-        return 1
+        # If querying multiple environments, display the env name
+        if hubble_args.debug:
+            print("-- [%s] --" % green(env['section'].value))
+
+        # If our invocation name is not 'hubble'
+        if not sys.argv[0].endswith('hubble'):
+            # Use the invocation name as our 'cmd'
+            env.add({'cmd': cmd_path(sys.argv[0], conf)})
+
+        if hubble_args.execute:
+            # Use the command provided
+            env.add({'cmd': hubble_args.execute})
+
+        # At this point we should know our 'cmd' to run
+        if 'cmd' not in env:
+            raise RuntimeError("Please specify a 'cmd' somewhere in "
+                               "your config")
+
+        # If --debug; print out our env config and pass along the
+        # --debug arg
+        if hubble_args.debug:
+            # For cinder client debug
+            if env['cmd'].endswith('cinder'):
+                env.add({'CINDERCLIENT_DEBUG': '1'})
+            print("%r\n" % env)
+            other_args.insert(0, '--debug')
+
+        # Grab a copy of the local environment and inject it into
+        # our environment
+        environ = os.environ.copy()
+        environ.update(env.to_dict())
+
+        try:
+            # Run the requested command
+            p = Popen([env['cmd'].value] + other_args,
+                      stdout=PIPE,
+                      stderr=PIPE,
+                      env=environ)
+            processes.append((p, env['section'].value))
+        except OSError as e:
+            if e.errno == 2:
+                print("-- No such executable '%s', you must specify the "
+                      "executable in the [hubble-commands] section of the "
+                      "config (See README)" % env['cmd'].value)
+            message = "exec failed '%s' - %s" % (env['cmd'].value, e)
+            raise RuntimeError(message)
+
+    for p, env in processes:
+        if len(environments) != 1:
+            print("-- [%s] --" % green(env))
+        # Wait for the command to complete
+        stdout, stderr = p.communicate()
+        sys.stdout.write(stdout.decode('utf-8'))
+        sys.stderr.write(stderr.decode('utf-8'))

--- a/hubble/shell.py
+++ b/hubble/shell.py
@@ -191,12 +191,8 @@ def get_environments(args, choice, config):
 
         # Populate environment vars by running opt-cmd
         # if -o was passed on the commandline
-        if args.option:
-            if 'opt-cmd' not in env:
-                log.warning("provided -o|--option, but 'opt-cmd' is not "
-                            "defined in '%s' section" % env['section'].value)
-            else:
-                env.add(run(env['opt-cmd'].value, env))
+        if 'opt-cmd' in env:
+            env.add(run(env['opt-cmd'].value, env))
 
         # Populate environment vars by running the env-cmd if it exists
         if 'env-cmd' in env:

--- a/hubble/shell.py
+++ b/hubble/shell.py
@@ -155,7 +155,7 @@ def get_cmd(argv, conf, env, hubble_args):
         return hubble_args.execute
 
     try:
-        return env['cmd']
+        return env['cmd'].value
     except KeyError:
         raise RuntimeError("Please specify a 'cmd' somewhere in "
                            "your config")
@@ -270,9 +270,8 @@ def execute_environment(cmd, env, hubble_args, other_args):
     # --debug arg
     if hubble_args.debug:
         # For cinder client debug
-        if cmd.value.endswith('cinder'):
+        if cmd.endswith('cinder'):
             env.add({'CINDERCLIENT_DEBUG': '1'})
-        env.add({'cmd': cmd.value})
         print("%r\n" % env)
         other_args.insert(0, '--debug')
 
@@ -283,7 +282,7 @@ def execute_environment(cmd, env, hubble_args, other_args):
 
     try:
         # Run the requested command
-        p = Popen([cmd.value] + other_args,
+        p = Popen([cmd] + other_args,
                   stdout=PIPE,
                   stderr=PIPE,
                   env=environ)

--- a/hubble/shell.py
+++ b/hubble/shell.py
@@ -146,6 +146,22 @@ def green(msg):
     return "\033[92m%s\033[0m" % msg
 
 
+def get_cmd(conf, env, hubble_args):
+    # If our invocation name is not 'hubble'
+    if not sys.argv[0].endswith('hubble'):
+        return cmd_path(sys.argv[0], conf)
+
+    if hubble_args.execute:
+        # Use the command provided
+        return hubble_args.execute
+
+    try:
+        return env['cmd']
+    except KeyError:
+        raise RuntimeError("Please specify a 'cmd' somewhere in "
+                           "your config")
+
+
 def get_environments(args, choice, config):
     """ Get the environment collection requested from args.env """
     sections = [choice]
@@ -172,6 +188,20 @@ def get_environments(args, choice, config):
             return "opt.%s" % i[0], str(i[1])
         # Add the args to the environment as opt.'<arg_name>'
         env.add(dict(map(f, vars(args).items())), section)
+
+        # Populate environment vars by running opt-cmd
+        # if -o was passed on the commandline
+        if args.option:
+            if 'opt-cmd' not in env:
+                log.warning("provided -o|--option, but 'opt-cmd' is not " \
+                            "defined in '%s' section" % env['section'].value)
+            else:
+                env.add(run(env['opt-cmd'].value, env))
+
+        # Populate environment vars by running the env-cmd if it exists
+        if 'env-cmd' in env:
+            env.add(run(env['env-cmd'].value, env))
+
         # Apply var expansion
         results.append(env.eval())
     return results
@@ -238,6 +268,40 @@ def eval_args(conf, parser):
     return (arg1, arg2, env)
 
 
+def execute_environment(conf, env, hubble_args, other_args):
+    cmd = get_cmd(conf, env, hubble_args)
+
+    # If --debug; print out our env config and pass along the
+    # --debug arg
+    if hubble_args.debug:
+        # For cinder client debug
+        if cmd.endswith('cinder'):
+            env.add({'CINDERCLIENT_DEBUG': '1'})
+        print("%r\n" % env)
+        other_args.insert(0, '--debug')
+
+    # Grab a copy of the local environment and inject it into
+    # our environment
+    environ = os.environ.copy()
+    environ.update(env.to_dict())
+
+    try:
+        # Run the requested command
+        p = Popen([env['cmd'].value] + other_args,
+                  stdout=PIPE,
+                  stderr=PIPE,
+                  env=environ)
+    except OSError as e:
+        if e.errno == 2:
+            print("-- No such executable '%s', you must specify the "
+                  "executable in the [hubble-commands] section of the "
+                  "config (See README)" % env['cmd'].value)
+        message = "exec failed '%s' - %s" % (env['cmd'].value, e)
+        raise RuntimeError(message)
+
+    return p
+
+
 def main():
     logging.basicConfig(format='-- %(message)s')
     log.setLevel(logging.CRITICAL)
@@ -292,65 +356,8 @@ def main():
     environments = get_environments(hubble_args, choice, conf)
     processes = []
     for env in environments:
-        # Populate environment vars by running opt-cmd
-        # if -o was passed on the commandline
-        if hubble_args.option:
-            if 'opt-cmd' not in env:
-                message = "provided -o|--option, but 'opt-cmd' is not " \
-                          "defined in '%s' section" % env['section'].value
-                raise RuntimeError(message)
-            env.add(run(env['opt-cmd'].value, env))
-
-        # Populate environment vars by running the env-cmd if it exists
-        if 'env-cmd' in env:
-            env.add(run(env['env-cmd'].value, env))
-
-        # If querying multiple environments, display the env name
-        if hubble_args.debug:
-            print("-- [%s] --" % green(env['section'].value))
-
-        # If our invocation name is not 'hubble'
-        if not sys.argv[0].endswith('hubble'):
-            # Use the invocation name as our 'cmd'
-            env.add({'cmd': cmd_path(sys.argv[0], conf)})
-
-        if hubble_args.execute:
-            # Use the command provided
-            env.add({'cmd': hubble_args.execute})
-
-        # At this point we should know our 'cmd' to run
-        if 'cmd' not in env:
-            raise RuntimeError("Please specify a 'cmd' somewhere in "
-                               "your config")
-
-        # If --debug; print out our env config and pass along the
-        # --debug arg
-        if hubble_args.debug:
-            # For cinder client debug
-            if env['cmd'].endswith('cinder'):
-                env.add({'CINDERCLIENT_DEBUG': '1'})
-            print("%r\n" % env)
-            other_args.insert(0, '--debug')
-
-        # Grab a copy of the local environment and inject it into
-        # our environment
-        environ = os.environ.copy()
-        environ.update(env.to_dict())
-
-        try:
-            # Run the requested command
-            p = Popen([env['cmd'].value] + other_args,
-                      stdout=PIPE,
-                      stderr=PIPE,
-                      env=environ)
-            processes.append((p, env['section'].value))
-        except OSError as e:
-            if e.errno == 2:
-                print("-- No such executable '%s', you must specify the "
-                      "executable in the [hubble-commands] section of the "
-                      "config (See README)" % env['cmd'].value)
-            message = "exec failed '%s' - %s" % (env['cmd'].value, e)
-            raise RuntimeError(message)
+        p = execute_environment(conf, env, hubble_args, other_args)
+        processes.append((p, env['section'].value))
 
     for p, env in processes:
         if len(environments) != 1:

--- a/hubble/tests/test_shell.py
+++ b/hubble/tests/test_shell.py
@@ -1,0 +1,73 @@
+#   Copyright 2012 Derrick J. Wippler
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+from hubble.shell import main
+
+from io import StringIO
+import tempfile
+import unittest
+import os
+
+
+def safe_eval(io_out):
+    str_out = io_out.getvalue()
+    print(str_out)
+    if len(str_out) != 0:
+        return eval(str_out)
+    return {}
+
+
+def run_main(config, argv):
+    stdout = StringIO()
+    stderr = StringIO()
+    ret = main(argv, stdout=stdout, stderr=stderr, files=[config])
+    return ret, safe_eval(stdout), safe_eval(stderr)
+
+
+def generate_test_cmd(print_stmt):
+    with tempfile.NamedTemporaryFile(delete=False) as fd:
+        fd.write("#! /usr/bin/env python\n"
+                 "from __future__ import print_function\n"
+                 "import os\n"
+                 "{0}\n".format(print_stmt).encode())
+        fd.flush()
+    os.chmod(fd.name, 0o755)
+    return fd.name
+
+
+class TestHubble(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.opt_cmd = generate_test_cmd("print('opt-injected-key=value\\n')")
+        cls.cmd = generate_test_cmd("print(dict(os.environ))")
+
+    @classmethod
+    def tearDownClass(cls):
+        os.unlink(cls.opt_cmd)
+        os.unlink(cls.cmd)
+
+    def test_opt_cmd(self):
+        config = StringIO(u"""[hubble]
+            name=Test
+            opt-cmd={0} ${{opt.option}} ${{name}}
+            [dfw]
+            cmd={1}
+            SOME=Thing""".format(self.opt_cmd, self.cmd))
+        config.name = "test.conf"
+
+        ret, out, err = run_main(config, ["hubble", "-o", "my-option",
+                                          "dfw", "--debug"])
+        self.assertEqual(ret, 0)
+        self.assertEqual(out['SOME'], "Thing")
+        self.assertEqual(out['opt-injected-key'], "value")

--- a/hubble/tests/test_shell.py
+++ b/hubble/tests/test_shell.py
@@ -12,12 +12,12 @@
 #   See the License for the specific language governing permissions and
 #   limitations under the License.
 
-from hubble.shell import main
-
 from io import StringIO
+import os
 import tempfile
 import unittest
-import os
+
+from hubble.shell import main
 
 
 def safe_eval(io_out):

--- a/hubble/tests/unit/test_hubble.py
+++ b/hubble/tests/unit/test_hubble.py
@@ -12,12 +12,12 @@
 #   See the License for the specific language governing permissions and
 #   limitations under the License.
 
-from hubble.shell import empty, Env, get_environments, run, to_dict
-from hubble.config import parse_configs
-
-from io import StringIO
 import argparse
+from io import StringIO
 import unittest
+
+from hubble.config import parse_configs
+from hubble.shell import empty, Env, get_environments, run, to_dict
 
 
 class TestEnv(unittest.TestCase):

--- a/hubble/tests/unit/test_hubble.py
+++ b/hubble/tests/unit/test_hubble.py
@@ -43,6 +43,7 @@ class TestEnv(unittest.TestCase):
         parser = argparse.ArgumentParser()
         parser.add_argument('env')
         parser.add_argument('--user', default='', required=False)
+        parser.add_argument('--option')
         args = parser.parse_args(['blah'])
 
         file = StringIO("[hubble]\n"

--- a/hubble/tests/unit/test_hubble.py
+++ b/hubble/tests/unit/test_hubble.py
@@ -12,13 +12,12 @@
 #   See the License for the specific language governing permissions and
 #   limitations under the License.
 
+from hubble.shell import empty, Env, get_environments, run, to_dict
+from hubble.config import parse_configs
+
+from io import StringIO
 import argparse
 import unittest
-
-from six.moves import StringIO
-
-from hubble.config import parse_configs
-from hubble.shell import empty, Env, get_environments, run, to_dict
 
 
 class TestEnv(unittest.TestCase):
@@ -46,7 +45,7 @@ class TestEnv(unittest.TestCase):
         parser.add_argument('--option')
         args = parser.parse_args(['blah'])
 
-        file = StringIO("[hubble]\n"
+        file = StringIO(u"[hubble]\n"
                         "name=My name is ${FIRST} ${last}\n"
                         "[prod-meta]\n"
                         "meta=['name', 'place']\n"

--- a/test-requirements.txt
+++ b/test-requirements.txt
@@ -7,3 +7,7 @@ pytest-cov>=2.3.0
 # flake8 plugins
 pep8-naming>=0.4.1
 flake8-import-order>=0.8
+
+# Tox
+tox>=2.3.1
+tox-pyenv>=1.0.3

--- a/tox.ini
+++ b/tox.ini
@@ -4,7 +4,7 @@
 # and then run "tox" from this directory.
 
 [tox]
-envlist = py27, py34, py35, flake8
+envlist = py27, py35, flake8
 skip_missing_interpreters = True
 
 [testenv]
@@ -29,7 +29,7 @@ commands =
     coveralls --verbose
 
 [flake8]
-ignore = E731
+ignore = E731, I100
 max-complexity = 10
 import-order-style = google
 application-import-names = hubble

--- a/tox.ini
+++ b/tox.ini
@@ -4,7 +4,7 @@
 # and then run "tox" from this directory.
 
 [tox]
-envlist = py27, py35, flake8
+envlist = py27, py34, py35, flake8
 skip_missing_interpreters = True
 
 [testenv]

--- a/tox.ini
+++ b/tox.ini
@@ -29,7 +29,7 @@ commands =
     coveralls --verbose
 
 [flake8]
-ignore = E731, I100
+ignore = E731
 max-complexity = 10
 import-order-style = google
 application-import-names = hubble

--- a/tox.ini
+++ b/tox.ini
@@ -30,6 +30,6 @@ commands =
 
 [flake8]
 ignore = E731
-#max-complexity = 10
+max-complexity = 10
 import-order-style = google
 application-import-names = hubble

